tools/fiptool/fiptool.c

```diff
@@ -0,0 +1,873 @@
+		if (memcmp(&toc_entry->uuid, &uuid_null, sizeof(uuid_t)) == 0) {
+			terminated = 1;
+			break;
+		}
+
+		/*
+		 * Build a new image out of the ToC entry and add it to the
+		 * table of images.
+		 */
+		image = malloc(sizeof(*image));
+		if (image == NULL)
+			log_err("malloc");
+
+		memcpy(&image->uuid, &toc_entry->uuid, sizeof(uuid_t));
+
+		image->buffer = malloc(toc_entry->size);
+		if (image->buffer == NULL)
+			log_err("malloc");
+
+		/* Overflow checks before memory copy. */
+		if (toc_entry->size > (uint64_t)-1 - toc_entry->offset_address)
+			log_errx("FIP %s is corrupted", filename);
+		if (toc_entry->size + toc_entry->offset_address > st.st_size)
+			log_errx("FIP %s is corrupted", filename);
+
+		memcpy(image->buffer, buf + toc_entry->offset_address,
+		    toc_entry->size);
+		image->size = toc_entry->size;
+
+		image->toc_entry = get_entry_lookup_from_uuid(&toc_entry->uuid);
+		if (image->toc_entry == NULL)
+			log_errx("Image with unknown UUID in FIP %s", filename);
+
+		assert(image->toc_entry->image == NULL);
+		/* Link backpointer from lookup entry. */
+		image->toc_entry->image = image;
+		add_image(image);
+
+		toc_entry++;
+	}
+
+	if (terminated == 0)
+		log_errx("FIP %s does not have a ToC terminator entry",
+		    filename);
+	free(buf);
+	return 0;
+}
+
+static image_t *read_image_from_file(toc_entry_t *toc_entry, char *filename)
+{
+	struct stat st;
+	image_t *image;
+	FILE *fp;
+
+	fp = fopen(filename, "r");
+	if (fp == NULL)
+		log_err("fopen %s", filename);
+
+	if (fstat(fileno(fp), &st) == -1)
+		log_errx("fstat %s", filename);
+
+	image = malloc(sizeof(*image));
+	if (image == NULL)
+		log_err("malloc");
+
+	memcpy(&image->uuid, &toc_entry->uuid, sizeof(uuid_t));
+
+	image->buffer = malloc(st.st_size);
+	if (image->buffer == NULL)
+		log_err("malloc");
+	if (fread(image->buffer, 1, st.st_size, fp) != st.st_size)
+		log_errx("Failed to read %s", filename);
+	image->size = st.st_size;
+	image->toc_entry = toc_entry;
+
+	fclose(fp);
+	return image;
+}
+
+static int write_image_to_file(image_t *image, char *filename)
+{
+	FILE *fp;
+
+	fp = fopen(filename, "w");
+	if (fp == NULL)
+		log_err("fopen");
+	if (fwrite(image->buffer, 1, image->size, fp) != image->size)
+		log_errx("Failed to write %s", filename);
+	fclose(fp);
+	return 0;
+}
+
+static int fill_opts(struct option *opts, int has_arg)
+{
+	int i;
+
+	/* -1 because the terminator entry is ignored. */
+	for (i = 0; i < toc_entries_len - 1; i++) {
+		opts[i].name = toc_entries[i].cmdline_name;
+		opts[i].has_arg = has_arg;
+		opts[i].flag = NULL;
+		opts[i].val = 0;
+	}
+	return i;
+}
+
+static void add_opt(struct option *opts, int idx, char *name,
+    int has_arg, int val)
+{
+	opts[idx].name = name;
+	opts[idx].has_arg = has_arg;
+	opts[idx].flag = NULL;
+	opts[idx].val = val;
+}
+
+static int info_cmd(int argc, char *argv[])
+{
+	image_t *image;
+	uint64_t image_offset;
+	uint64_t image_size = 0;
+	fip_toc_header_t toc_header;
+	int i;
+
+	if (argc != 2)
```

Comment:
Hi @afaerber

> Very specifically I am commenting on the above hard check on argc == 2, whereas pack_cmd() further below checks for argc < 2 and has getopt handling.

Yes, if support were needed for options on the info sub-command, that would need changing. I'm sure we'd be open to that.

> Further, main() only has rudimentary -v/--verbose handling.

fip_create doesn't have that option at all. We can always build on this.

> fip_create by contrast had all options handled in a central place, so I was pointing out that this refactoring seems to be making the use case of adding options harder here. Can be fixed as follow-up of course.

Processing options centrally makes it more difficult for the sub-command to handle unsupported/incompatible options for that sub-command. This way, it's clearer which options apply to which sub-command. There could be better code sharing in the handling of options that apply to multiple sub-commands, and the mapping of sub-commands to valid options could be represented more centrally in the code (e.g. via some data structure). I agree this can be done later if needed, but the current set of options / sub-commands don't really justify it.

> ... Therefore I would like to package one official, maintained version of fip_create / fiptool and use that to combine vendor's ATF (where not yet upstream) and our U-Boot...

Sounds good to me.

> So while you are right that vendors can fork and patch your software, I would really like to have one official tool with suitable command line options. 

I never said that; I agree it's much better if vendors contribute their changes back upstream into 1 official tool.

>Be it --pad=0x4000 or --target=amlogic/gxbb or anything like that, defaulting to current vanilla behavior. It would be a global option then, applying to all commands - either after the subcommand or before it.

Although those commands sounds reasonable and would apply to multiple sub-commands, I'm not sure they're truly global. Anyway, that's just an implementation detail.

> Writing local patches is not the problem - how to get them mainline is, to not have to maintain patches for dozens of vendors myself. And the commit message here specifically promises a better base for future features, which I'm seeing this as. A uniform SBBR world is not what we're seeing in practice, so we'll need to handle variances. [EDIT: I realize that I misread TBBR as SBBR... still a valid point.]

Hmm, it sounds like you're at the hard end of handling platform variation! Up to now, we haven't been very aware of platform specific requirements like this. We'd be grateful of any code contributions or raising of requirements on our [issue tracker](https://github.com/arm-software/tf-issues/issues).

> Restricting bugfix contributions for a tool that does not directly touch on vendor IP is counter-productive. Would you seriously rather have us fork the tool into U-Boot (like Amlogic/Hardkernel did) or as a separate project to get around your contribution hurdles? That can't be in your best interest, unless you plan to drop it here then.

I'm not trying to introduce any contribution hurdles for you; I'm just saying what the current process is. Of course I'd rather avoid forks, but this is the first time I heard there was a fork, or that the the TF contribution process is causing a problem.

If required I can go back to our legal team and try to improve the current process, but I can't unilaterally change it. Are you only concerned about bug fix changes or contributing more generally (including feature enhancements)? Are the CLA terms an issue for you, or the fact that there is one at all? We can take this offline if you want.